### PR TITLE
Cookbook Segment Deprecation

### PR DESCRIPTION
**Pull request by lamont-granquist**
The elimination of cookbook segments from the server protocol that has been discussed for awhile now.

### CODE REVIEW

**Comment by lamont-granquist** [comment]
Some noodling on the problem was here:

https://github.com/chef/chef_objects/pull/79

Somewhere there's a discussion between me + @danielsdeleo where he suggested how the backcompat should work.  Now that we have API versioning this should work even better (and might work even better than I've outlined here, I haven't looked at the details of that).

I don't like `all_files` very much but `files` would be my first choice and that top-level key is taken by the `files` segment already.  Maybe it'd be safe to leverage API versioning harder in order to 'break' this key (?) but I'd want someone who knew more about the implications of that to lead me down that road...

**Comment by jaymzh** [comment]
How does this affect lazy loading?

**Comment by coderanger** [comment]
Not at all, the actual loading code doesn't really use segments, it uses direct path globs like you would expect. Other than the file sync code and the templates/cookbook_files system, I don't think much really cares about segments as is.

**Comment by jaymzh** [comment]
Well that's what I"m talking about the - the templates / cookbook_files system part of Chef. If I have a template, chef-client (if I've told it to) will lazy-load that source of the template only if it ends up needing it. You (for reasons I can't fathom) turned this option off in Chef 12, but we turned it back on (cause performance)...

**Comment by danielsdeleo** [comment]
Nothing about this would make lazy loading impossible, there would just be a slightly different implementation internally (i.e., you would know that files are templates because the first element of the path is "templates").

As for the server's versioning capability, we're in a crappy spot where we can't use it to its full potential yet. This is because 11.x servers are still supported by 12.x clients, but 11.x servers are totally unaware of versioning. Therefore, they are incapable of rejecting requests for API versions that they don't support.

I don't remember exactly the details of what we decided for the compatibility behavior. Probably we need to think through the compat matrix of:

* server supports non-segments: (t/f)
* uploading client supports non-segments: (t/f)
* downloading client supports non-segments: (t/f)

**Comment by jaymzh** [comment]
OK so it sounds like it's possible but will require some work to make work in the new model. I just want to make sure that's taken into account. Our performance *tanked* when lazy loading became optional, (and we put a bunch of stuff on hosts that didn't need it... ), so that has to stick around.

**Comment by lamont-granquist** [comment]
Yeah, I wasn't planning on taking that away at all.  The way it works now is that you take all the files in the manifest and nuke the `files` and `templates` segments and use the resultant list to synchronize, its pretty straightforwards to just remove paths starting with `files/` and `templates/` instead.  We can call that out, but it just isn't going to be a change, so I didn't include it in the RFC.

**Comment by jaymzh** [comment]
Cool. That makes sense. We should call it out just so it doesn't get dropped (since the default doesn't do this anymore, I can see someone not thinking about it when they go to implement it).

**Comment by danielsdeleo** [comment]
The code would be exactly like the code that knife uses to categorize files into segments now, I would expect.

**Comment by jaymzh** [comment]
@danielsdeleo  I'm not worried about knife... :)

**Comment by lamont-granquist** [comment]
Okay so @danielsdeleo it sounds like I probably wrote this the correct way then.  Since we use a new 'all_files' segment we don't change the semantics of `files` and don't break the back-compat world there, and then the RFC is written dependent upon what can be determined on either side of the connection, and I called out the behavior that should occur if you can't determine the protocol to use as well.  Hopefully its also clear that back-compat takes priority over the letter of what is written in this RFC if it turns out we got any of it wrong.

I didn't want to go into details of the implementation that we'd talked about but one of the ideas was to use the initial sending of the cookbook files up to the server to determine the server version so that the PUT of the metadata at the end of the upload could pick the correct format to use.  We could use that, or we could use API versioning, etc.  I think that can be handled when writing the code though and in PR review...

**Comment by thommay** [comment]
:+1:

**Comment by btm** [comment]
:+1:

File: new/cookbook-segment-deprecation.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+new 'segment' and accept it since the Chef Server currently does strict enforcement of file
+segments server-side.  This will also generate a version bump of the Chef Server protocol which
+will be used by clients to determine if the Chef Server supports this field or not.
+
+New knife clients that support this segment will use the Chef Server protocol version to determine
+if they MUST upload cookbooks using the old segment format, or if they SHOULD upload using only
+the `all_files` format.
+
+For backwards compatibility, the Chef Server MUST serve cookbooks with both standard segments and the
+new `all_files` segment unless it can determine that the client prefers one or other.  If it can
```
Comment (stevendanna) [inline new/cookbook-segment-deprecation.md]: I believe we should be able to use the API version here as well.  For versions before this was introduced or no-version we return the current response.  For versions after it was introduce, serve the new response.  That is, I'm unclear on the cases where we would be unable to determine client preference.

What should the server do with all_files members that don't map to the old segments?

**Comment by stevendanna** [comment]
Tentatively :+1: but I think for new API responses, the RFC should include example payloads and example returns.  Also, we need to specify what the version 0 or version 1 API response will look like when a cookbook was uploaded using all_files that includes segments that don't map to the existing segments.

File: new/cookbook-segment-deprecation.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+be determined then the server MUST serve cookbooks to old clients with the old format, and SHOULD
+serve cookbooks to new clients with only the `all_files` segment.
+
+New Chef Clients MUST accept either old style segments, cookbooks with the new `all_files` segment
+or cookbook metadata with both styles (and SHOULD favor the `all_files` segment).
+
+This will achieve backwards compatibility and the ability for knife, chef-client and the chef-server
```
Comment (stevendanna) [inline new/cookbook-segment-deprecation.md]: Doesn't need to be in the RFC, but we need to make sure that cheffs style commands are also compatible with this (we've broken those on past API changes) as well as widely used tools such as knife-backup and knife-ec-backup.

File: new/cookbook-segment-deprecation.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+# Title
+
+Cookbook Segment Deprecation
+
+## Motivation
+
+    As a Chef User/Developer,
+    I want to be able to extend the structure of cookbooks without fighting with a wire protocol
+    written 6 years ago,
```
Comment (stevendanna) [inline new/cookbook-segment-deprecation.md]: I lol'd but I'm not sure this is the best motivation description :).  How about,

> As a Chef User/Developer
> I want to be able to extend the structure of cookbooks easily,
> in order to quickly adapt to new automation needs.

**Comment by smurawski** [comment]
:+1:

File: new/cookbook-segment-deprecation.md (hunk 1 above)
Comment (danielsdeleo) [inline new/cookbook-segment-deprecation.md]: > I believe we should be able to use the API version here as well.
I could go either way. What we have to do is support a matrix of old client/new server, and new client/old server, and we won't be able to 100% rely on versioning in the server until at least Chef 13.

> What should the server do with all_files members that don't map to the old segments?
You're gonna need a client and server that support "segmentless" cookbooks to take advantage.

File: new/cookbook-segment-deprecation.md (hunk 2 above)
Comment (lamont-granquist) [inline new/cookbook-segment-deprecation.md]: I'm hoping that the implementation is at a deep enough layer in the objects that client code won't have to be updated everywhere.

File: new/cookbook-segment-deprecation.md (hunk 3 above)
Comment (lamont-granquist) [inline new/cookbook-segment-deprecation.md]: Yeah, I find writing motivation descriptions completely demotivating and useless...

File: new/cookbook-segment-deprecation.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+This RFC will add a special `all_files` segment to cookbooks which will duplicate (initially)
+all of the existing cookbook manifest along with arbitrary files outside of the current scope of
+any of the cookbook segments.  The Chef Server will need to be patched in order to support this
+new 'segment' and accept it since the Chef Server currently does strict enforcement of file
+segments server-side.  This will also generate a version bump of the Chef Server protocol which
+will be used by servers to determine if the Chef Client supports this field or not.  Clients MAY
+implement code which determines the acceptable API version of the Server and then only send
+the new API requests to the server.  Clients MAY instead send requests which include both the
+old segment and new segment data.  The Server MUST detect the Client's desired API version and
```
Comment (lamont-granquist) [inline new/cookbook-segment-deprecation.md]: So I'm fairly certain that what I just wrote is broken.  If clients send up a cookbook version with 'all_files' set in addition to the other segments then on an old server it'll just barf.  That suggests to me that the RFC should really state that there is no 'mixed mode' and that servers will always know what the client supports so should send that format, and that clients MUST figure out if they're talking to an old server or not (we have no generic mechanism for this, so I can't suggest an implementation here, but for cookbook uploads it is simple and the client uploads all the sandbox files first and can determine the API version from those responses).  Then once the client determines the server API version it should send the appropriate requests.

Comment (lamont-granquist) [inline new/cookbook-segment-deprecation.md]: ^ that comment is out of date, but i'm still seeing it on my GH page even though I've pushed a complete rewrite.

**Comment by btm** [comment]
:+1:

**Comment by thommay** [comment]
Approved, @chef/rfc-editors do your thing

**Pull request closed without merging**